### PR TITLE
Image resizer

### LINKED ISSUE #5167: Implement the image resize handler
The functionality of the [enhanced image plugin in v4](https://ckeditor.com/cke4/addon/image2) is missing in v5.

image

A follow-up of https://github.com/ckeditor/ckeditor5/issues/1330.

---

If you'd like to see this feature implemented, add 👍 to this post.

### PR DESCRIPTION
**Pull request by mlewand**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced image widget resizer. Closes ckeditor/ckeditor5#5167.

---

### Additional information

* Requires ckeditor/ckeditor5-widget#94.
* PoC of image resizer.
* Unit tests might fail for the time being as I did not add test coverage yet.

### CODE REVIEW

**Comment by Mgsy** [comment]
Resize doesn't work if there are images in neighbouring table cells. 

bug_cke5

**Comment by Mgsy** [comment]
The editor crashes after pasting content during resizing.

## Steps to reproduce
1. Start resizing.
1. During resizing press <kbd>Ctrl</kbd> + <kbd>C</kbd>, then <kbd>Ctrl</kbd> + <kbd>V</kbd>.

## Current result
The editor crashes.

## Errors
```
imageediting.js:71 Uncaught TypeError: Cannot read property 'getAttribute' of undefined
    at Object.isCentered (imageediting.js:71)
    at vu.updateSize (resizertopbound.js:64)
    at xu.updateSize (resizecontext.js:189)
    at dr.n (widgetresizer.js:70)
    at dr.fire (emittermixin.js:207)
    at HTMLDocument.n (emittermixin.js:235)

writer.js:1481 Uncaught TypeError: Cannot read property 'getAttribute' of undefined
    at bc (writer.js:1481)
    at pc.setAttribute (writer.js:370)
    at Object.callback (resizecontext.js:157)
    at Gc._runPendingChanges (model.js:738)
    at Gc.change (model.js:160)
    at xu.commit (resizecontext.js:156)
    at dr.<anonymous> (widgetresizer.js:109)
    at dr.fire (emittermixin.js:207)
    at HTMLDocument.n (emittermixin.js:235)
```

## GIF
bug_cke5

File: src/image.js [hunk 1]
```diff
@@ -34,7 +35,7 @@ export default class Image extends Plugin {
 	 * @inheritDoc
 	 */
 	static get requires() {
-		return [ ImageEditing, Widget, ImageTextAlternative ];
+		return [ ImageEditing, Widget, ImageTextAlternative, ImageResize ];
```
Comment (Reinmar) [inline src/image.js]: Just like `ImageCaption` or `ImageStyle`, this should be an optional plugin. So, it should not be listed here.

File: tests/image.js [hunk 2]
```diff
@@ -66,8 +66,9 @@ describe( 'Image', () => {
 			setModelData( model, '[<image alt="alt text" src="/assets/sample.png"></image>]' );
 
 			expect( getViewData( view ) ).to.equal(
-				'[<figure class="ck-widget ck-widget_selected image" contenteditable="false">' +
+				'[<figure class="ck-widget ck-widget_selected ck-widget_with-resizer image" contenteditable="false">' +
```
Comment (Reinmar) [inline tests/image.js]: All those changes in this file can be reverted after you change https://github.com/ckeditor/ckeditor5-image/pull/298/files#r312947942.

**Comment by Reinmar** [comment]
- Tests
    - [x]  [ML] Update tests after our recent changes.
    - [x]  [ML] Test with responsive images markup.
    - [x]  [ML] Verify that `srcsetAttributeConverter()` doesn't override our width.
- Collaboration examples
    - [x]  [PK] Secure resizing for image disappearing during resizing
    - [x]  [PK] Use command
    - [x]  [PK] Disabling the command should disable the UI
- Misc
    - [x]  [ML] Percentage values
    - [x]  Fix the table selection handler
    - [x]  [PK] Fix the img{width} being float and resizer width being int, which adds this ugly 
    - [x]  [ML] Make sure building API docs doesn't throw; document the config option and the `ImageResize` class
    - [x]  [PK] Feature guide
    - [x]  [KT] Repeate testing
    - [x]  [KT] Make sure other browsers are tested
- Followups
    - Handle ESC key ckeditor/ckeditor5-widget#98
    - Split ImageResize to ImageResizeEditing and ImageResizeUI ckeditor/ckeditor5-image#304
    - Improve resize cursor so that it is not overridden during the resize. ckeditor/ckeditor5-widget#100
    - Better detect centered images ckeditor/ckeditor5-widget#102
    - Review other bugs reported by Kacper here. 
    - Complete API docs.
    - Rewrite automated tests (100% CC and proper tests for classes). ckeditor/ckeditor5-widget#101

**Comment by Reinmar** [comment]
Rounding done. It indeed fixed the visual issue.

**Comment by Reinmar** [comment]
OK, it seems that we've got it. There's a set of followups, but we can merge the functionality which is definitely already functional.

**Pull request merged**